Task 0

melgenek opened the pull request.

Savotii committed:
Реализация UserDAOImpl

Savotii committed:
Добавление класса помощника и получение коннекта к mysql

Savotii committed:
Некоторые настройки бинов

Savotii committed:
Некоторые настройки бинов

melgenek commented on src/main/java/com/andersen/spring/dao/AbstractDAO.java:
```diff
@@ -0,0 +1,27 @@
+
+public abstract class AbstractDAO {
+
+    private static long id;
```
AtomicLong. Веб - параллельная среда

melgenek commented on src/main/java/com/andersen/spring/storage/MarketStorage.java:
```diff
@@ -0,0 +1,91 @@
+
+    }
+
+    private void initFileProducts(File file) {
```
Не надо такое делать.
Просто файлик под версию добавь

melgenek commented on src/main/java/com/andersen/spring/storage/MarketStorage.java:
```diff
@@ -0,0 +1,91 @@
+
+            for (String line : lines) {
+              Product product = gson.fromJson(line, Product.class);
+              products.put(product, 10);
```
количество - лишнее

Savotii replied:
Переделал на лист

melgenek commented on src/main/java/App.java:
```diff
@@ -1,5 +1,25 @@
+
+import java.sql.Connection;
+
 public class App {
```
Этот класс, наверняка, лишний

melgenek commented on src/main/resources/application-context.xml:
```diff
@@ -0,0 +1,41 @@
+    <bean id="App" class="App"/>
+
+    <!-- MySqlHelper beans -->
+    <bean id="MySqlConnection" class="com.andersen.spring.jdbc.MySqlHelper" factory-method="createConnection"/>
```
Соединение так не создается. Оно на каждый запрос новое (ну или из пула)

melgenek commented on src/main/java/com/andersen/spring/facade/MarketFacadeImpl.java:
```diff
@@ -0,0 +1,83 @@
+    public MarketFacadeImpl(Connection connection) {
+
+        this.connection = connection;
+        this.productServiceImpl = new ProductServiceImpl(connection);
```
за тебя Spring создает классы

melgenek commented on src/main/java/com/andersen/spring/dao/AbstractDAO.java:
```diff
@@ -0,0 +1,27 @@
+
+import com.andersen.spring.storage.MarketStorage;
+
+public abstract class AbstractDAO {
```
Generic лучше.
AbsractDao<T>

Savotii committed:
Донастройка файла бинов.
Попытка написания теста.

Savotii committed:
Попытка написания теста.

Savotii committed:
Попытка написания теста.

Savotii committed:
Попытка написания теста.

Savotii committed:
почистил от лишнего

Savotii committed:
почистил от лишнего

melgenek commented on src/main/java/com/andersen/spring/facade/MarketFacadeImpl.java:
```diff
@@ -0,0 +1,83 @@
+
+    public MarketFacadeImpl(ProductServiceImpl productServiceImpl, UserServiceImpl userServiceImpl) {
+
+        this.connection = MySqlHelper.createConnection();
```
за рамки dao соединение на выходит

Savotii replied:
исправлено

melgenek commented on src/main/java/com/andersen/spring/impl/ProductDAOImpl.java:
```diff
@@ -0,0 +1,237 @@
+        List<Product> products = new LinkedList<Product>();
+
+        try {
+            PreparedStatement statement = connection.prepareStatement(GET_BY_TITLE_QUERY);
```
соединение получаешь на своем хелпере перед выполнением, а потом закрываешь

melgenek commented on src/main/java/com/andersen/spring/impl/ProductServiceImpl.java:
```diff
@@ -0,0 +1,47 @@
+    private ProductDAOImpl productDAO;
+
+    public ProductServiceImpl() {//Connection connection) {
+        this.productDAO = new ProductDAOImpl();
```
все создания объектов делает спринг

Savotii replied:
Исправлено

melgenek commented on src/main/java/com/andersen/spring/impl/ProductDAOImpl.java:
```diff
@@ -0,0 +1,237 @@
+import java.util.LinkedList;
+import java.util.List;
+
+public class ProductDAOImpl extends AbstractDAO implements ProductDAO {
```
extends AbstractDAO<Product>

Savotii replied:
Исправлено

melgenek commented on src/main/java/com/andersen/spring/jdbc/MySqlHelper.java:
```diff
@@ -0,0 +1,45 @@
+import java.sql.DriverManager;
+import java.sql.SQLException;
+
+public class MySqlHelper {
```
Не должно быть статических методов.  Создай класс и сделай его полем в dao. И инициализируй его спрингом

Savotii replied:
Исправлено

melgenek commented on src/main/java/com/andersen/spring/storage/MarketStorage.java:
```diff
@@ -0,0 +1,101 @@
+import java.util.List;
+import java.util.Map;
+
+public class MarketStorage {
```
удали, раз не используешь

Savotii replied:
удалил.

melgenek commented on src/test/java/MyTest.java:
```diff
@@ -0,0 +1,44 @@
+    @Test
+    public void TestServices() {
+
+        Product product = mock(Product.class);
```
уже есть такой мок

Savotii replied:
Убрал

melgenek commented on src/main/java/com/andersen/spring/impl/ProductDAOImpl.java:
```diff
@@ -0,0 +1,237 @@
+        return pr;
+    }
+
+    public Object createdToMock(Object item) {
```
????

Savotii replied:
Убрал.

melgenek commented on src/test/java/MyTest.java:
```diff
@@ -0,0 +1,44 @@
+import org.mockito.runners.MockitoJUnitRunner;
+
+@RunWith(MockitoJUnitRunner.class)
+public class MyTest {
```
Что тестируем? Надо сервисы тестировать

Savotii replied:
Убрал

melgenek commented on src/test/java/MyTest.java:
```diff
@@ -0,0 +1,44 @@
+
+        when(productDAO.createdToMock(product)).thenReturn(product);
+
+        Product p = (Product)productDAO.createdToMock(product);
```
????

Savotii replied:
убрал

Savotii committed:
Поправил комменты, дописал хелпер к mysql протестировал создание и удаление записей.

Savotii committed:
Откорректировал тест

melgenek commented on src/main/java/com/andersen/spring/dao/UserDAO.java:
```diff
@@ -0,0 +1,13 @@
+
+import java.util.List;
+
+public interface UserDAO extends DAO {
```
Dao < User >

Savotii replied:
поправил.

melgenek commented on src/main/java/com/andersen/spring/facade/MarketFacadeImpl.java:
```diff
@@ -0,0 +1,80 @@
+    }
+
+    public User updateUser(User user) {
+        return null;
```
?

Savotii replied:
проглядел, видимо.

melgenek commented on src/main/java/com/andersen/spring/facade/MarketFacadeImpl.java:
```diff
@@ -0,0 +1,80 @@
+
+    public boolean deleteUser(long id) {
+
+        boolean action = false;
```
я думаю, что это не нужно

Savotii replied:
Согласен, лишнее.

melgenek commented on src/main/java/com/andersen/spring/impl/ProductServiceImpl.java:
```diff
@@ -0,0 +1,51 @@
+
+public class ProductServiceImpl implements ProductService {
+
+     private ProductDAOImpl productDAO;
```
пиши абстракции, реализация будет известна спрингу.
Замени ProductDAOImpl на Dao< Product >

Savotii replied:
сложный вопрос, т.к. в ProductDao есть больше возможностей. Интерфейс DAO определяет 4 метода или 5, а конкретные интерфейсы имеют несколько расширенный функционал, поэтому к DAO будет проблематично.

melgenek replied:
Тогда ProductDao, но не Impl

melgenek commented on src/main/java/com/andersen/spring/impl/UserDAOImpl.java:
```diff
@@ -0,0 +1,191 @@
+            e.printStackTrace();
+        }
+
+        helper.closeConnection();
```
finally{
  .close()
}

Savotii replied:
поправил.

melgenek commented on src/main/java/com/andersen/spring/impl/UserServiceImpl.java:
```diff
@@ -0,0 +1,45 @@
+
+public class UserServiceImpl implements UserService {
+
+    private UserDAOImpl userDAO;
```
Dao<  User >

melgenek commented on src/main/java/com/andersen/spring/jdbc/H2Helper.java:
```diff
@@ -0,0 +1,24 @@
+import java.sql.DriverManager;
+import java.sql.SQLException;
+
+public class H2Helper {
```
не нужен, кажется

Savotii replied:
Убрал

melgenek commented on src/main/resources/application-context.xml:
```diff
@@ -0,0 +1,53 @@
+
+    <bean id="Product" class="com.andersen.spring.entity.Product"/>
+
+    <bean id="User" class="com.andersen.spring.entity.User" />
```
Разберись с bean scopes. http://proselyte.net/tutorials/spring-tutorial-full-version/bean-scope/
Скорее всего бины, которые только содержат поля и никакой логики, можно создавать через new Product. 
Или через prototype+lookup-method, если, к примеру, у продукта задано поле по умолчанию цена=10$
В твоем случае один объект на все прилоение.

melgenek commented on src/main/java/com/andersen/spring/dao/AbstractDAO.java:
```diff
@@ -0,0 +1,24 @@
+
+import java.sql.Connection;
+
+public abstract class AbstractDAO {
```
не нужен

melgenek commented on src/main/java/com/andersen/spring/controllers/AbstractService.java:
```diff
@@ -0,0 +1,5 @@
+package com.andersen.spring.controllers;
+
+public abstract class AbstractService {
```
не нужен

Savotii replied:
Убрал.

melgenek commented on src/test/java/MyTest.java:
```diff
@@ -0,0 +1,53 @@
+    ProductServiceImpl productServiceImpl;
+
+    @InjectMocks
+    UserServiceImpl userServiceImpl;
```
Посредством моков ты обрезаешь зависимости. 
Так, если ты тестируешь фасад, то он зависит от сервисов, которые в свою очередь зависят от дао. 
Но если ты фасад будет создан используя моковые сервисы, то зависимости на дао уже не будет - мы вложенность зависимостей ограничим одним уровнем. Для этого и нужны моки

melgenek commented on src/test/java/MyTest.java:
```diff
@@ -0,0 +1,53 @@
+    @Mock
+    User user;
+
+    @InjectMocks
```
Обычно делают один класс - один тест.
ProductServiceTest
UserServiceTest и т.д.

melgenek commented on src/test/java/MyTest.java:
```diff
@@ -0,0 +1,53 @@
+    @InjectMocks
+    UserServiceImpl userServiceImpl;
+
+    @InjectMocks
```
Это должен быть мок, ведь у тебя нет "хелпера".

melgenek commented on src/test/java/MyTest.java:
```diff
@@ -0,0 +1,53 @@
+    @Test
+    public void TestServices() {
+
+        UserService userServiceImpl = mock(UserServiceImpl.class);
```
уже создал полями

Savotii replied:
Убрал.

melgenek commented on src/test/java/MyTest.java:
```diff
@@ -0,0 +1,53 @@
+        ProductService productServiceImpl = mock(ProductServiceImpl.class);
+
+        productDAO = mock(ProductDAOImpl.class);
+        when( productServiceImpl.create(product)).thenReturn(product);
```
when( productDao.create(product)).thenReturn(product);

Savotii replied:
Запутался, мы сервис тестим или дао?

melgenek commented on src/test/java/MyTest.java:
```diff
@@ -0,0 +1,53 @@
+        productDAO = mock(ProductDAOImpl.class);
+        when( productServiceImpl.create(product)).thenReturn(product);
+
+        userDAO = mock(UserDAOImpl.class);
```
уже создан

melgenek commented on src/test/java/MyTest.java:
```diff
@@ -0,0 +1,53 @@
+        when( productServiceImpl.create(product)).thenReturn(product);
+
+        userDAO = mock(UserDAOImpl.class);
+        when(userServiceImpl.create(user)).thenReturn(user);
```
Ты должен проверить хоть что-то. Всегда есть assert на реально работающем коде, не на моке. 
Объект проверки не должен быть моком, или жеты проверишь, что вернул то, что сказал вернуть и в вызов метода мы даже не войдем

melgenek commented on src/test/java/MyTest.java:
```diff
@@ -0,0 +1,53 @@
+import org.mockito.InjectMocks;
+import org.mockito.Mock;
+
+public class MyTest {
```
@RunWith(MockitoJUnitRunner.class)

Savotii replied:
Вернул.

Savotii committed:
Отработка некоторой части комментариев.

Savotii committed:
Убрал AbstractDAO,  в ProductDAOImp и user... имплементил соответствующие дао. чтобы спринг съел их правильно.
Тест пока не пересматривал

Savotii committed:
Переопределение equals, написание теста по новой.
За изучение брал http://qaru.site/questions/35009/difference-between-mock-and-injectmocks

Savotii committed:
UserService test

Savotii committed:
UserService test

Savotii committed:
Попытка выродить lookup-method , как я понял.

melgenek commented on src/main/java/com/andersen/spring/facade/MarketFacadeImpl.java:
```diff
@@ -0,0 +1,73 @@
+
+public class MarketFacadeImpl implements MarketFacade {
+
+    private ProductServiceImpl productServiceImpl;
```
Используй интерфейсы. Классы ничего не должны знать о реализациях

Savotii replied:
Переделал.

melgenek commented on src/test/java/UserServiceTest.java:
```diff
@@ -0,0 +1,45 @@
+    public void testService()
+    {
+
+        when(userService.create(user)).thenReturn(new User("test", "email.test", "+380000000"));
```
один вызов метода проверяемого класса - один тест

melgenek commented on src/test/java/UserServiceTest.java:
```diff
@@ -0,0 +1,45 @@
+        when(userService.create(user)).thenReturn(new User("test", "email.test", "+380000000"));
+        assertEquals(new User("test", "email.test", "+380000000"), userService.create(user));
+
+        when(userService.update(user)).thenReturn(new User("test", "email.test", "+3811111111"));
```
Ты ничего не проверяешь. У тебя выполение программы не заходит в класс. Мокай только внутренности класса, а не сам класс

melgenek commented on src/main/java/com/andersen/spring/impl/ProductServiceImpl.java:
```diff
@@ -0,0 +1,54 @@
+    }
+
+    public Product getById(long id) {
+        return (Product) productDAO.getById(id);
```
приведения типов не нужны

Savotii committed:
Попытка выродить lookup-method , как я понял.

Savotii committed:
Обработка комментариев

Savotii committed:
Реализация списочного теста

Savotii committed:
Реализация списочного теста

Savotii committed:
Реализация списочного теста

Savotii committed:
Merge branch 'master' into task_0

# Conflicts:
#	src/main/java/App.java
